### PR TITLE
Staking pools fetching

### PR DESCRIPTION
**Pull request by saneery**
*[Set blockscout to work with staking app in POSDAO network](https://github.com/poanetwork/blockscout/issues/1709)*

This PR for fetching staking pools from POSDAO network.

I added:
* **Fetcher.StakingPools** - fetches current list of pools from the network on per block.
* **Import.Runner.Validators** - inserts new pools list to the database.

Required variables:
* POS_VALIDATORS_CONTRACT
* POS_STAKING_CONTRACT

To run fetcher you need to enable it in the `indexer/config/config.exs`:
```
config :indexer, Indexer.Fetcher.StakingPools.Supervisor, disabled?: false
```
Also you need to disable BlockReward supervisor
```
config :indexer, Indexer.Fetcher.BlockReward.Supervisor, disabled?: true
```

### CODE REVIEW

File: apps/explorer/lib/explorer/staking/pools_reader.ex [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      }
+    else
+      _ ->
+        :error
+    end
+  end
+
+  defp call_staking_method(method, params) do
+    %{^method => resp} =
+      Reader.query_contract(config(:staking_contract_address), abi("staking.json"), %{
```
Comment (zachdaniel) [inline apps/explorer/lib/explorer/staking/pools_reader.ex]: Might be good to put these in module attributes since they are constants.

**Comment by zachdaniel** [comment]
I approved just based on the code itself, but I'm not really familiar with DApps themselves, so others will definitely want to weigh in on that aspect.

File: apps/explorer/lib/explorer/staking/pools_reader.ex (hunk 1 above)
Comment (saneery) [reply apps/explorer/lib/explorer/staking/pools_reader.ex]: Are you about the contract address?

Comment (zachdaniel) [reply apps/explorer/lib/explorer/staking/pools_reader.ex]: No, the abi file names.

File: apps/explorer/lib/explorer/staking/pools_reader.ex [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  def pool_data(staking_address) do
+    with {:ok, [mining_address]} <- call_validators_method("miningByStakingAddress", [staking_address]),
+         {:ok, [is_active]} <- call_staking_method("isPoolActive", [staking_address]),
+         {:ok, [delegator_addresses]} <- call_staking_method("poolDelegators", [staking_address]),
+         delegators_count = Enum.count(delegator_addresses),
+         {:ok, [staked_amount]} <- call_staking_method("stakeAmountTotalMinusOrderedWithdraw", [staking_address]),
+         {:ok, [is_validator]} <- call_validators_method("isValidator", [mining_address]),
+         {:ok, [was_validator_count]} <- call_validators_method("validatorCounter", [mining_address]),
+         {:ok, [is_banned]} <- call_validators_method("isValidatorBanned", [mining_address]),
+         {:ok, [banned_unitil]} <- call_validators_method("bannedUntil", [mining_address]),
```
Comment (goodsoft) [inline apps/explorer/lib/explorer/staking/pools_reader.ex]: Typo

File: apps/indexer/lib/indexer/fetcher/staking_pools.ex [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  defp changeset(%{staking_address: staking_address} = pool) do
+    {:ok, mining_address} = Chain.Hash.Address.cast(pool[:mining_address])
+
+    data =
+      pool
+      |> Map.delete(:staking_address)
+      |> Map.put(:mining_address, mining_address)
+
+    %{
+      name: "anonymous",
```
[the remark this answers is not in the document]
Comment (saneery) [reply apps/indexer/lib/indexer/fetcher/staking_pools.ex]: In staking app pools are all anonymous

**Comment by saneery** [comment]
@goodsoft 

> From what I see, you make `2 + 9 * N` (where `N` is total count of pools) of `eth_call` requests once on the launch, and then after _every_ new block import. Method calls can and should be batched together to reduce request overhead.

Are there examples of how to batch?

> Also, I'm not sure of how POSDAO works, but is it really necessary to query _each and every_ existing pool after _each and every_ incoming block? Is it maybe possible to gather a list of updated pools from
> block content?

No, from block content only able to get few events but they aren't enough

**Comment by goodsoft** [comment]
> Are there examples of how to batch?

Currently only token balance requests get batched, see `apps/explorer/lib/explorer/token/balance_reader.ex`. Ideally, all other smart contract queries should be done in the same way, if they happen often enough.

The basic idea is to use `Explorer.SmartContract.Reader.query_contracts` instead of `query_contract`.

> No, from block content only able to get few events but they aren't enough

Sad. Even more reason to batch requests as much as possible.

File: apps/indexer/lib/indexer/fetcher/staking_pools.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (goodsoft) [reply apps/indexer/lib/indexer/fetcher/staking_pools.ex]: Okay, I assumed there is a way to manually specify a name for certain well-known addresses.

**Comment by saneery** [comment]
@goodsoft I've added batching requests. Could you look at this?

**Comment by vbaranov** [comment]
>@vbaranov we'll need to compare app server and node performance before and after deploying this PR. If the amount of requests will cause too much of a burden, we'll need to improve batching.

@goodsoft I suppose it shouldn't affect any instance (except POSDAO) as pools fetcher will be switched off by default. Meanwhile, sure, I will check performance on POSDAO instance when we will be ready to deploy it.

**Pull request merged**